controllers/datafoundationclient.go

```diff
@@ -123,6 +133,23 @@ func (r *dataFoundationClientReconciler) parseSpec(offering *v1alpha1.ManagedFus
 
 func (r *dataFoundationClientReconciler) reconcilePhases() (ctrl.Result, error) {
 	if !r.offering.DeletionTimestamp.IsZero() {
+		if found, err := r.hasOCSVolumes(); err != nil {
+			return ctrl.Result{}, err
+		} else if found {
+			r.Log.Info("PVs provisioned by ODF  storage classes exists, cannot proceed with uninstallation")
+			return ctrl.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
+		}
+		r.Log.Info("issuing a delete for Default StorageClassClaims")
+		if err := r.delete(&r.defaultBlockStorageClassClaim); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to issue a delete for defaultBlockStorageClassClaim: %v", err)
+		}
+		if err := r.delete(&r.defaultFSStorageClassClaim); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to issue a delete for defaultFSStorageClassClaim: %v", err)
+		}
```

Comment:
Shouldn't we delete non-default storageClassClaims as well that are provisioned by `cephfs.csi.ceph.com` or `rbd.csi.ceph.com`?

Reply:
That is a very good question! @ezio-auditore what do you think?

Reply:
What I think
_If the user creates his custom storageClassClaim using the the default storageclient spec only then we need to delete those storageClassClaims,else we need to ignore it._

@nb-ohad @bindrad If this sounds ok I will implement the same here.

Reply:
> default storageclient spec

Are we even allowing customer to create non-default storageClient? Or do we have plans to allow customer to create non-default storageClients?

Reply:
> If the user creates his custom storageClassClaim using the the default storageclient spec
@ezio-auditore  that does not make any sense. The reason someone will create another claim is if they want some aspects of the storageclass to be different

Reply:
@nb-ohad 
So why do we even need to remove the storageclassclaims created by users?

Reply:
Because we are uninstalling the controller that reconciles it. That is precisely the gap that Dhruv has highlighted. 
But on the other hand, the CRD is not getting uninstalled and the CR themselves are not owned by our controller. 
So it is a gray area. But then again, we do remove the default storage class claims, so the behavior is not consistent.

I am not sure what are is best practice here

Reply:
Do you think it would be beneficial to prevent uninstallation until all user-created storage class claims have been deleted?

Reply:
According to me, the customers will be creating additional storageClassClaims to utilize storageClass with different settings. The expectation from the customer's side will be to have the storage running seamlessly. So we should block the uninstallation of the DFC offering if there are user-created storageClassClaims on the cluster.

@Madhu-1 @nb-ohad WDYT?

Reply:
Sounds good to me. This is the same as blocking the uninstall or reqeueing the request if there are still PV in the cluster.

Reply:
The existence of storageclassclaims or even storageclasses does not infer usage of storage. I am against blocking uninstallation based on that

Reply:
@Madhu-1 Will the storageClassClaims get removed if the storageClient CR/client-operator is removed?

Reply:
If there are any PVCs utilizing the storageclass of SCC, it cannot be removed. Similarly, the StorageClient cannot be deleted until all the SCCs using it are deleted.

Reply:
I forgot to post my comment on this one :facepalm: 

Instead of checking PV, it would be better to check SCC before deleting the storageClient for two reasons. Firstly, we are creating StorageClient and StorageClassClaims, so it makes sense to check for the resources we have created. Secondly, listing PVs can be a heavy operation compared to listing SCCs, as there are very few.

Reply:
@Madhu-1 We must check for PVs, that is not even a question. We need to be sure that no-one is consuming any storage from our provider before we delete the client. We should not care how the storageclass, backing the PV, was created (manually or via  SCC). 

The question here was about SCC, I think we should just delete the lot of them

Reply:
@nb-ohad, the ocs-client-operator is already checking for any PV before deleting the SCC. Why do we need to recheck the PV here again?

>The question here was about SCC, I think we should just delete the lot of them

You mean list all the SCC, delete them, and wait for them to be deleted before deleting the storage client right?

Reply:
Do we support Users creating custom storageclass and static provisioning without SCC in managed service?

Reply:
> Do we support Users creating custom storageclass and static provisioning without SCC in managed service?

On consumer clusters, customer can do anything they want to.

Reply:
> You mean list all the SCC, delete them, and wait for them to be deleted before deleting the storage client right?

I think we can issue a delete for all SCCs and then issue a delete for storageClient. No need to wait for the removal of all SCCs before deleting storageClient because client-operator won't delete storageClient unless all SCCs are removed.

Reply:
@Madhu-1 I am aware the ocs-client-operator is already checking for that. 
But a k8s delete operation is unrevertable so we check for potential lock conditions, from the operator that owns the lifecycle of the offering (in this case the client operator) before issuing the delete request.